homeassistant/auth/mfa_modules/__init__.py

```diff
@@ -90,6 +97,40 @@ def setup_schema(self) -> Optional[vol.Schema]:
         raise NotImplementedError
 
 
+class SetupFlow(data_entry_flow.FlowHandler):
+    """Handler for the setup flow."""
+
+    def __init__(self, auth_module: MultiFactorAuthModule,
+                 user_id: str) -> None:
+        """Initialize the setup flow."""
+        self._auth_module = auth_module
+        self._user_id = user_id
+
+    async def async_step_init(
+            self, user_input: Optional[Dict[str, str]] = None) \
+            -> Dict[str, Any]:
+        """Handle the first step of setup flow.
+
+        Return self.async_show_form(step_id='init') if user_input == None.
+        Return await self.async_finish(flow_result) if finish.
+        """
+        errors = {}  # type: Dict[str, str]
+
+        if user_input:
+            result = await self._auth_module.async_setup_user(
+                self._user_id, user_input)
+            return self.async_create_entry(
+                title=self._auth_module.name,
+                data={'result': result}
+            )
+
+        return self.async_show_form(
+            step_id='init',
+            data_schema=self._auth_module.setup_schema,
```

Comment:
I think that all MFA modules need to have their own flows instead of a generic flow wrapped around a `setup_schema` property. Because in the case of TOTP, the flow will need to be:

- Start login flow
- Login flow generates a secret and stores it on instance of login flow
- Returns show_form with a QR code in the description (injected as base64 via `description_placeholder`)
- User scans code and enters a code to verify it scanned it correctly
- TOTP module is enabled for user

This can only be done if each MFA module has their own setup flow.

Example of how GitHub does this:

![image](https://user-images.githubusercontent.com/1444314/44520860-7c626900-a6d1-11e8-848d-6aeb3adbe71f.png)

Reply:
Generic SetupFlow is default implement, other module can implement its own if need.

Reply:
Okay, in that case we should not have it be the default method, because now we are requiring an extra property `setup_schema` to be optionally defined.

Instead, if an MFA module wants to use the default config flow, they can just do:

```python
setup_schema = vol.Schema({})
return SetupFlow(self.name, setup_schema, user_id)
```

That way we keep the API for MFA modules simpler while providing clean API that is also similar to Auth Providers.

Reply:
Done